Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp

```diff
@@ -1117,6 +1117,26 @@ class ConstantFoldingPhase : public Phase {
 
                 break;
             }
+
+            case HasStructureWithFlags: {
+                const AbstractValue& abstractValue = m_state.forNode(node->child1());
+                unsigned flags = node->structureFlags();
+                ASSERT(flags);
+
+                if (abstractValue.m_type && !(abstractValue.m_type & ~SpecCell) && abstractValue.m_structure.isFinite()) {
+                    bool noStructureWithAnyFlag = true;
+                    abstractValue.m_structure.forEach([&] (RegisteredStructure structure) {
+                        if (structure->hasAnyOfBitFieldFlags(flags))
+                            noStructureWithAnyFlag = false;
+                    });
+                    if (noStructureWithAnyFlag) {
+                        m_graph.convertToConstant(node, jsBoolean(false));
+                        changed = true;
+                    }
+                }
```

Comment:
Actually, this code is a bit wrong since Structure can change their flags without transitions. For example, `hasBeenFlattenedBefore` flag can be set without transitions. `didWatchReplacement` too. So this code is not correct since its flag is too generic. Some flags cannot be set without transition, but some are not. But right now, this code strongly assuming that these flags will not be modified.
I think you need to allow this folding only for specific Structure flags.

Reply:
Oh wow that's a great catch, introduced `Structure::bitFieldFlagsCantBeChangedWithoutTransition` to filter out those flags and even conditionally brought back removed clobberize based on it.